File: Chess-ios/Chess/asd/Board.swift
//
//  Board.swift
//  chess-program
//
//  Created by Cseri Ádám on 2022. 05. 02..
//

import Foundation

class Board {
    let board: Array<Array<Spot>>
    
    lazy var side: String = ""
    lazy var enPas: String = ""
    lazy var castPerm: String = ""
    lazy var halfMove: Int = 0
    lazy var fullMove: Int = 0
    
    init() {
        board = (1...8).map { _ in (1...8).map { _ in Spot() } }
        parseFEN()
    }
    
    func toPrint() {
        for row in board {
            let rank = row[0].id[1..<2]
            let spotsInARow = row.map{ it in it.piece.char }.joined(separator: "  ")
            
            print(rank, spotsInARow, separator: "   ")
        }
        let files = board[0].map{ it in it.id[0..<1] }.joined(separator: "  ")
        print("\n   ", files, "\n")
        
        print("""
Side: \(side)
En Passant: \(enPas)
Castling Permission: \(castPerm)
""")
    }
    
    func toPrintIDs() {
        for row in board {
            for spot in row {
                print(spot.id, terminator: " ")
            }
            print()
        }
    }
    
    func parseFEN(customStringFEN: String = START_FEN) {
        let customFEN = customStringFEN.components(separatedBy: " ")
        var rowIterator = 0
        var colIterator = 0
        
        for char in customFEN[0] {
            switch char {
            case "P":
                board[rowIterator][colIterator].piece = Pawn(isWhite: true)
            case "p":
                board[rowIterator][colIterator].piece = Pawn(isWhite: false)
            case "R":
                board[rowIterator][colIterator].piece = Rook(isWhite: true)
            case "r":
                board[rowIterator][colIterator].piece = Rook(isWhite: false)
            case "N":
                board[rowIterator][colIterator].piece = Knight(isWhite: true)
            case "n":
                board[rowIterator][colIterator].piece = Knight(isWhite: false)
            case "B":
                board[rowIterator][colIterator].piece = Bishop(isWhite: true)
            case "b":
                board[rowIterator][colIterator].piece = Bishop(isWhite: false)
            case "K":
                board[rowIterator][colIterator].piece = King(isWhite: true)
            case "k":
                board[rowIterator][colIterator].piece = King(isWhite: false)
            case "Q":
                board[rowIterator][colIterator].piece = Queen(isWhite: true)
            case "q":
                board[rowIterator][colIterator].piece = Queen(isWhite: false)
            case "/":
                rowIterator+=1
                colIterator = -1
            default:
                colIterator += Int(String(char))!-1
            }
            colIterator+=1
        }
        
        self.side = customFEN[1]
        self.castPerm = customFEN[2]
        self.enPas = customFEN[3]
        self.halfMove = Int(customFEN[4])!
        self.fullMove = Int(customFEN[5])!
    }
    
    func makeMove(move: Move) {
        var capture = type(of: getSpotById(id: move.to).piece) != type(of: Empty())
        
        let movingPiece = getSpotById(id: move.from).piece
        getSpotById(id: move.from).piece = Empty()
        getSpotById(id: move.to).piece = movingPiece
        
        if isEnPasMove(toSquare: move.to, piece: movingPiece) {
            capture = true
            captureEnPas(move: move)
        }
        
        if isPromotionMove(move: move) {
            promotePawn(spot: getSpotById(id: move.to), promotion: move.promotion!)
        }
        
        if isCastlingMove(move: move) {
            replaceRook(move: move)
        }
        
        side = (side == "w") ? "b" : "w"
        enPas = updateEnPas(piece: movingPiece, move: move)
        castPerm = updateCastPerm(castPerm: castPerm)
        halfMove = updateHalfMove(piece: movingPiece, capture: capture)
        fullMove = updateFullMove()
    }
    
    func replaceRook(move: Move) {
        if (move.player.isWhite && move.from == "e1" && move.to == "c1" && castPerm.contains("Q")) {
            let rookTemp = board[7][0].piece
            board[7][0].piece = Empty()
            board[7][3].piece = rookTemp
        }
        if (move.player.isWhite && move.from == "e1" && move.to == "g1" && castPerm.contains("K")) {
            let rookTemp = board[7][7].piece
            board[7][7].piece = Empty()
            board[7][5].piece = rookTemp
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "c8" && castPerm.contains("q")) {
            let rookTemp = board[0][0].piece
            board[0][0].piece = Empty()
            board[0][3].piece = rookTemp
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "g8" && castPerm.contains("k")) {
            let rookTemp = board[0][7].piece
            board[0][7].piece = Empty()
            board[0][5].piece = rookTemp
        }
    }
    
    func isCastlingMove(move: Move) -> Bool {
        if castPerm.contains("-") {
            return false
        }
        if (move.player.isWhite && move.from == "e1" && move.to == "c1" && castPerm.contains("Q")) {
            return true
        }
        if (move.player.isWhite && move.from == "e1" && move.to == "g1" && castPerm.contains("K")) {
            return true
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "c8" && castPerm.contains("q")) {
            return true
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "g8" && castPerm.contains("k")) {
            return true
        }
        return false
    }
    
    func isPromotionMove(move: Move) -> Bool {
        return move.promotion != nil
    }
    
    func isEnPasMove(toSquare: String, piece: Piece) -> Bool {
        return toSquare == enPas && type(of: piece) == type(of: Pawn())
    }
    
    func promotePawn(spot: Spot, promotion: String) {
        if (promotion == "q") {
            spot.piece = (side == "w") ? Queen(isWhite: true) : Queen(isWhite: false)
        }
        if (promotion == "r") {
            spot.piece = (side == "w") ? Rook(isWhite: true) : Rook(isWhite: false)
        }
        if (promotion == "n") {
            spot.piece = (side == "w") ? Knight(isWhite: true) : Knight(isWhite: false)
        }
        if (promotion == "b") {
            spot.piece = (side == "w") ? Bishop(isWhite: true) : Bishop(isWhite: false)
        }
    }
    
    func captureEnPas(move: Move) {
        let direction = (move.player.isWhite) ? 1 : -1
        let (rank, file) = getPosition(piece: getSpotById(id: move.to).piece)
        board[rank+direction][file].piece = Empty()
    }
    
    func getSpotById(id: String) -> Spot {
        return board.flatMap { $0 }.first { $0.id == id }!
    }
    
    func getPosition(piece: Piece) -> (Int, Int) {
        let row = board.first { row in row.contains { $0.piece === piece } }!
        let rankIndex = board.findIndexNS(customArray: row)
        let fileIndex = board[rankIndex].firstIndex { $0.piece === piece }!
        return (rankIndex, fileIndex)
    }
    
    func filterNotSafeMoves(from: String, moveList: Array<Spot>) -> Array<Spot> {
        return moveList.filter { toSpot in
            let to = toSpot.id
            let saveFromPiece = getSpotById(id: from).piece
            let savetoPiece = getSpotById(id: to).piece
            
            getSpotById(id: to).piece = getSpotById(id: from).piece
            getSpotById(id: from).piece = Empty()
            
            let enemySpotList = allSpotsOfPlayer(isWhite: !getSpotById(id: to).piece.isWhite!)
            let kingUnderAttack = isKingUnderCheck(enemySpotList: enemySpotList)
            
            getSpotById(id: from).piece = saveFromPiece
            getSpotById(id: to).piece = savetoPiece
            
            return !kingUnderAttack
        }
    }
    
    func isKingUnderCheck(enemySpotList: Array<Spot>) -> Bool {
        let array = enemySpotList.map { $0.piece.possibleCaptureMoves(board: self).map { inner$0 in type(of: inner$0.piece) } }.flatMap { $0 }
        return Piece.contains(array: array, elem: King())
    }
    
    func allSpotsOfPlayer(isWhite: Bool) -> Array<Spot> {
        return board.flatMap { rank in rank.filter { type(of: $0.piece) != type(of: Empty()) && isWhite == $0.piece.isWhite } }
    }
    
    func updateEnPas(piece: Piece, move: Move) -> String {
        if (type(of: piece) != type(of: Pawn())) {
            return "-"
        }
        if (piece.isWhite! && board[6].map { $0.id }.contains(move.from) && board[4].map { $0.id }.contains(move.to)) {
            return String(move.from.first!) + String(Int(String(move.from.last!))!+1)
        }
        if (!piece.isWhite! && board[1].map { $0.id }.contains(move.from) && board[3].map { $0.id }.contains(move.to)) {
            return String(move.from.first!) + String(Int(String(move.from.last!))!-1)
        }
        return "-"
    }
    
    func updateCastPerm(castPerm: String) -> String {
        if (castPerm == "-") {
            return castPerm
        }
        
        var castPerm2 = (type(of: board[7][0].piece) != type(of: Rook())) ? castPerm.filter { $0 != "Q" } : castPerm
        castPerm2 = (type(of: board[7][7].piece) != type(of: Rook())) ? castPerm2.filter { $0 != "K" } : castPerm2
        castPerm2 = (type(of: board[7][4].piece) != type(of: King())) ? castPerm2.filter { $0 != "K" && $0 != "Q" } : castPerm2
        castPerm2 = (type(of: board[0][0].piece) != type(of: Rook())) ? castPerm.filter { $0 != "q" } : castPerm2
        castPerm2 = (type(of: board[0][7].piece) != type(of: Rook())) ? castPerm2.filter { $0 != "k" } : castPerm2
        castPerm2 = (type(of: board[0][4].piece) != type(of: King())) ? castPerm2.filter { $0 != "k" && $0 != "q" } : castPerm2
        
        if castPerm2.isEmpty {
            castPerm2 = "-"
        }
        return castPerm2
    }
    
    func updateHalfMove(piece: Piece, capture: Bool) -> Int {
        if (type(of: piece) == type(of: Pawn()) || capture) {
            return 0
        }
        return self.halfMove+1
    }
    
    func updateFullMove() -> Int {
        return (side == "w") ? self.fullMove+1 : self.fullMove
    }
}

File: Chess-android/app/src/main/kotlin/shared_code/Board.kt
package shared_code

class Board {
    val board = Array(RANKS) { Array(FILES) { Spot() } }

    lateinit var side: String
    lateinit var enPas: String
    lateinit var castPerm: String
    lateinit var halfMove: Number
    lateinit var fullMove: Number

    /**
     * Setting up the board manually
     */
    init {
        parseFEN()
    }

    /**
     * Prints the board.
     */
    fun toPrint() {
        print("\n\n")
        board.forEach { row ->
            val ranks = row[0].id[1]  // contains the ranks (numbers)
            val spotsInARow = row.joinToString("  ") { spot -> spot.piece.char.toString() } // contains 1 row of spots and pieces

            println("""
                $ranks   $spotsInARow
            """.trimIndent())
        }
        val files = board[0].joinToString("  ") { it.id[0].toString() }  // contains the files (letters)
        println("\n    $files\n")

        println("""
            Side: $side
            En Passant: $enPas
            Castling Permission: $castPerm
        """.trimIndent())
    }

    fun toPrintIDs() {
        board.forEach { row ->
            row.forEach { spot ->
                print("${spot.id} ")
            }
            println()
        }
    }

    /**
     * Parses the string FEN to an actual board position.
     *
     *  - 1st field:
     *    - It has 7 slashes
     *    - It has exactly 8 fields
     *
     *  - 2nd field:
     *    - It has 1 character, either 'w' or 'b'
     *
     *  - 3rd field:
     *    - It can have 5 different characters, which can appear only once in this order: 'KQkq-'
     *
     *  - 4th field:
     *    - There is either a square id (i.e. 'c6') or a hyphen
     *
     *  - 5th field:
     *    - A number, between 0 and 50
     *    - It must reset to 0 if a piece is captured or a pawn advanced.
     *
     *  - 6th field:
     *    - a non-negative number, incremented by 1 after the black's shared_code.move
     */
    fun parseFEN(customStringFEN: String = START_FEN) {
        val customFEN = customStringFEN.split(" ")
        var rowIterator = 0
        var colIterator = 0

        customFEN[0].forEach { char ->
            when (char) {
                'P' -> board[rowIterator][colIterator].piece = Pawn(true)
                'p' -> board[rowIterator][colIterator].piece = Pawn(false)
                'R' -> board[rowIterator][colIterator].piece = Rook(true)
                'r' -> board[rowIterator][colIterator].piece = Rook(false)
                'N' -> board[rowIterator][colIterator].piece = Knight(true)
                'n' -> board[rowIterator][colIterator].piece = Knight(false)
                'B' -> board[rowIterator][colIterator].piece = Bishop(true)
                'b' -> board[rowIterator][colIterator].piece = Bishop(false)
                'K' -> board[rowIterator][colIterator].piece = King(true)
                'k' -> board[rowIterator][colIterator].piece = King(false)
                'Q' -> board[rowIterator][colIterator].piece = Queen(true)
                'q' -> board[rowIterator][colIterator].piece = Queen(false)
                '/' -> { rowIterator++; colIterator=-1 }
                else -> colIterator+=char.digitToInt()-1
            }
            colIterator++
        }

        side = customFEN[1]  /** Adding the current side */
        castPerm = customFEN[2]  /** Adding Castling Permissions */
        enPas = customFEN[3]  /** Adding En Passant square if available */
        halfMove = customFEN[4].toInt()
        fullMove = customFEN[5].toInt()
    }

    /**
     * Executes the player's shared_code.move and updates the board accordingly.
     */
    fun makeMove(move: Move) {
        // Decides if it's a capture shared_code.move
        var capture = getSpotById(move.to).piece::class != Empty::class

        // Get the piece from the 'shared_code.move.from' square. Saving it in a temporary variable
        val movingPiece = getSpotById(move.from).piece
        // Clear the spot
        getSpotById(move.from).piece = Empty()
        // shared_code.Move the piece to 'to' spot
        getSpotById(move.to).piece = movingPiece

        // If it's an En Passant capture shared_code.move, the other player's piece is set to empty
        if (isEnPasMove(move.to, movingPiece)) {
            capture = true
            captureEnPas(move)
        }
        // If it's a promotion shared_code.move, replace the current pawn with the selected piece
        if (isPromotionMove(move)) {
            promotePawn(getSpotById(move.to), move.promotion!!)
        }
        // If it's a castling shared_code.move, replace the rook to the correct position
        if (isCastlingMove(move)) {
            replaceRook(move)
        }

        // Updates the board data
        side = if (side == "w") "b" else "w"
        enPas = updateEnPas(movingPiece, move)
        castPerm = updateCastPerm(castPerm)
        halfMove = updateHalfMove(movingPiece, capture)
        fullMove = updateFullMove()
    }

    /**
     * If it's a castling shared_code.move, replaces the rook based on where the shared_code.King made the castling shared_code.move.
     */
    fun replaceRook(move: Move) {
        if (move.player.isWhite && move.from == "e1" && move.to == "c1" && 'Q' in castPerm) {
            val rookTemp = board[7][0].piece
            board[7][0].piece = Empty()
            board[7][3].piece = rookTemp
        }
        if (move.player.isWhite && move.from == "e1" && move.to == "g1" && 'K' in castPerm) {
            val rookTemp = board[7][7].piece
            board[7][7].piece = Empty()
            board[7][5].piece = rookTemp
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "c8" && 'q' in castPerm) {
            val rookTemp = board[0][0].piece
            board[0][0].piece = Empty()
            board[0][3].piece = rookTemp
        }
        if (!move.player.isWhite && move.from == "e8" && move.to == "g8" && 'k' in castPerm) {
            val rookTemp = board[0][7].piece
            board[0][7].piece = Empty()
            board[0][5].piece = rookTemp
        }
    }

    /**
     * Checks if the shared_code.move is a castling shared_code.move.
     */
    fun isCastlingMove(move: Move): Boolean {
        if ('-' in castPerm) return false
        if (move.player.isWhite && move.from == "e1" && move.to == "c1" && 'Q' in castPerm) return true
        if (move.player.isWhite && move.from == "e1" && move.to == "g1" && 'K' in castPerm) return true
        if (!move.player.isWhite && move.from == "e8" && move.to == "c8" && 'q' in castPerm) return true
        if (!move.player.isWhite && move.from == "e8" && move.to == "g8" && 'k' in castPerm) return true
        return false
    }

    /**
     * Checks if the shared_code.move is promotion shared_code.move
     */
    fun isPromotionMove(move: Move): Boolean {
        return move.promotion != null
    }

    /**
     * Replaces the promoted pawn with the selected piece.
     */
    fun promotePawn(spot: Spot, promotion: String) {
        if (promotion == "q") spot.piece = if (side == "w") Queen(true) else Queen(false)
        if (promotion == "r") spot.piece = if (side == "w") Rook(true) else Rook(false)
        if (promotion == "n") spot.piece = if (side == "w") Knight(true) else Knight(false)
        if (promotion == "b") spot.piece = if (side == "w") Bishop(true) else Bishop(false)
    }

    /**
     * Decides if the shared_code.move was an En Passant capture shared_code.move.
     */
    fun isEnPasMove(toSquare: String, piece: Piece): Boolean {
        return toSquare == enPas && piece::class == Pawn::class
    }

    /**
     * The function removes the opponent's piece from an En Passant shared_code.move
     */
    fun captureEnPas(move: Move) {
        val direction = if (move.player.isWhite) 1 else -1
        val (rank, file) = getPosition(getSpotById(move.to).piece)
        board[rank+direction][file].piece = Empty()
    }

    /**
     * After the black moved it increments itself by 1.
     */
    fun updateFullMove(): Number {
        if (side == "w") return this.fullMove.inc()
        return this.fullMove
    }

    /**
     * Resets half moves if it's a pawn shared_code.move or a piece is captured, otherwise it's value is incremented by 1.
     */
    fun updateHalfMove(piece: Piece, capture: Boolean): Number {
        if (piece::class == Pawn::class || capture) return 0
        return this.halfMove.inc()
    }

    /**
     * Checks if the Castling Permissions are still active.
     * The rules: https://www.flyordie.com/games/help/chess/en/games_rules_chess.html#castling
     */
    fun updateCastPerm(castPerm: String): String {
        if (castPerm == "-") return castPerm
        var castPerm2 = if (board[7][0].piece::class != Rook::class) castPerm.filterNot { it == 'Q' } else castPerm
        castPerm2 = if (board[7][7].piece::class != Rook::class) castPerm2.filterNot { it == 'K' } else castPerm2
        castPerm2 = if (board[7][4].piece::class != King::class) castPerm2.filterNot { it == 'K' || it == 'Q' } else castPerm2
        castPerm2 = if (board[0][0].piece::class != Rook::class) castPerm2.filterNot { it == 'q' } else castPerm2
        castPerm2 = if (board[0][7].piece::class != Rook::class) castPerm2.filterNot { it == 'k' } else castPerm2
        castPerm2 = if (board[0][4].piece::class != King::class) castPerm2.filterNot { it == 'k' || it == 'q' } else castPerm2
        if (castPerm2.isEmpty()) castPerm2 = "-"
        return castPerm2
    }

    /**
     * Checks if there is an En Passant square
     */
    fun updateEnPas(piece: Piece, move: Move): String {
        if (piece::class != Pawn::class) return "-"
        if (piece.isWhite!! && move.from in board[6].map { it.id } && move.to in board[4].map { it.id })
            return move.from.toMutableList()[0] + move.from.toMutableList()[1].inc().toString()
        if (!piece.isWhite && move.from in board[1].map { it.id } && move.to in board[3].map { it.id })
            return move.from.toMutableList()[0] + move.from.toMutableList()[1].dec().toString()
        return "-"
    }

    /**
     * Returns the spot by its ID.
     */
    fun getSpotById(id: String): Spot {
        return board.flatten().find { it.id == id }!!
    }

    /**
     * Returns the piece's current position on the matrix. It finds in which rank and file the piece is.
     */
    fun getPosition(piece: Piece): Pair<Int, Int> {
        val rankIndex = board.indexOf(
                this.board
                    .find { rank -> rank.find { spot -> spot.piece == piece } != null }!!
            )
        val fileIndex = board[rankIndex].indexOf(
                this.board[rankIndex]
                    .find { spot -> spot.piece == piece }
            )
        return Pair(rankIndex, fileIndex)
    }

    /**
     * A move is not safe if after that the King would be in check.
     * Checks if the selected spot is under attack by the other side.
     *   - saving the piece on the spot
     *   - moving the king to the selected spot
     *   - checking if the king is in check now
     *   - resetting the table
     */
    fun filterNotSafeMoves(from: String, moveList: List<Spot>): List<Spot> {
        return moveList.filter { toSpot ->
            val to = toSpot.id
            // saving the piece on the spot
            val saveFromPiece = getSpotById(from).piece
            val saveToPiece = getSpotById(to).piece

            // moving the king to the selected spot
            getSpotById(to).piece = getSpotById(from).piece
            getSpotById(from).piece = Empty()

            // checking if the king is in check now
            val enemySpotList = allSpotsOfPlayer(!getSpotById(to).piece.isWhite!!)
            val kingUnderAttack = isKingUnderCheck(enemySpotList)

            // resetting the table
            getSpotById(from).piece = saveFromPiece
            getSpotById(to).piece = saveToPiece

            return@filter !kingUnderAttack
        }
    }

    /**
     * Checks if the shared_code.King is under attack.
     */
    fun isKingUnderCheck(enemySpotList: List<Spot>): Boolean {
        return King::class in enemySpotList.map { it.piece.possibleCaptureMoves(this).map { innerIt -> innerIt.piece::class } }.flatten()
    }

    /**
     * The function collects all the pieces of a player.
     */
    fun allSpotsOfPlayer(isWhite: Boolean): List<Spot> {
        return board.map { rank ->
            rank.filter { it.piece::class != Empty::class && isWhite == it.piece.isWhite }
        }.filterNot { it == emptyList<Spot>() }.flatten()
    }
}


